In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import matplotlib.pyplot as plt # for data visualization purposes
import seaborn as sns # for statistical data visualization
%matplotlib inline
!pip install --upgrade category_encoders

In [2]:
data = 'adult.csv'

df = pd.read_csv(data, header=None)
df.drop(0,axis=0,inplace=True)
df.head()

0          1       2             3   4                   5   \
1  25    Private  226802          11th   7       Never-married   
2  38    Private   89814       HS-grad   9  Married-civ-spouse   
3  28  Local-gov  336951    Assoc-acdm  12  Married-civ-spouse   
4  44    Private  160323  Some-college  10  Married-civ-spouse   
5  18          ?  103497  Some-college  10       Never-married   

                  6          7      8       9     10 11  12             13  \
1  Machine-op-inspct  Own-child  Black    Male     0  0  40  United-States   
2    Farming-fishing    Husband  White    Male     0  0  50  United-States   
3    Protective-serv    Husband  White    Male     0  0  40  United-States   
4  Machine-op-inspct    Husband  Black    Male  7688  0  40  United-States   
5                  ?  Own-child  White  Female     0  0  30  United-States   

      14  
1  <=50K  
2  <=50K  
3   >50K  
4   >50K  
5  <=50K

In [3]:
col_names = ['age', 'workclass', 'fnlwgt', 'education', 'education_num', 'marital_status', 'occupation', 'relationship',
             'race', 'sex', 'capital_gain', 'capital_loss', 'hours_per_week', 'native_country', 'income']

df.columns = col_names
df.head()

age  workclass  fnlwgt     education education_num      marital_status  \
1  25    Private  226802          11th             7       Never-married   
2  38    Private   89814       HS-grad             9  Married-civ-spouse   
3  28  Local-gov  336951    Assoc-acdm            12  Married-civ-spouse   
4  44    Private  160323  Some-college            10  Married-civ-spouse   
5  18          ?  103497  Some-college            10       Never-married   

          occupation relationship   race     sex capital_gain capital_loss  \
1  Machine-op-inspct    Own-child  Black    Male            0            0   
2    Farming-fishing      Husband  White    Male            0            0   
3    Protective-serv      Husband  White    Male            0            0   
4  Machine-op-inspct      Husband  Black    Male         7688            0   
5                  ?    Own-child  White  Female            0            0   

  hours_per_week native_country income  
1             40  United-States  <=50K  
2             50  United-States  <=50K  
3             40  United-States   >50K  
4             40  United-States   >50K  
5             30  United-States  <=50K

In [4]:
# view summary of dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48842 entries, 1 to 48842
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             48842 non-null  object
 1   workclass       48842 non-null  object
 2   fnlwgt          48842 non-null  object
 3   education       48842 non-null  object
 4   education_num   48842 non-null  object
 5   marital_status  48842 non-null  object
 6   occupation      48842 non-null  object
 7   relationship    48842 non-null  object
 8   race            48842 non-null  object
 9   sex             48842 non-null  object
 10  capital_gain    48842 non-null  object
 11  capital_loss    48842 non-null  object
 12  hours_per_week  48842 non-null  object
 13  native_country  48842 non-null  object
 14  income          48842 non-null  object
dtypes: object(15)
memory usage: 6.0+ MB


In [5]:
# find categorical variables

categorical = [var for var in df.columns if df[var].dtype=='O']

print('There are {} categorical variables\n'.format(len(categorical)))

print('The categorical variables are :\n\n', categorical)

There are 15 categorical variables

The categorical variables are :

 ['age', 'workclass', 'fnlwgt', 'education', 'education_num', 'marital_status', 'occupation', 'relationship', 'race', 'sex', 'capital_gain', 'capital_loss', 'hours_per_week', 'native_country', 'income']


In [6]:
# check missing values in categorical variables

df[categorical].isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64

In [7]:
# view frequency counts of values in categorical variables

for var in categorical: 
    
    print(df[var].value_counts())

36    1348
35    1337
33    1335
23    1329
31    1325
      ... 
88       6
85       5
87       3
89       2
86       1
Name: age, Length: 74, dtype: int64
Private             33906
Self-emp-not-inc     3862
Local-gov            3136
?                    2799
State-gov            1981
Self-emp-inc         1695
Federal-gov          1432
Without-pay            21
Never-worked           10
Name: workclass, dtype: int64
203488    21
120277    19
190290    19
126569    18
125892    18
          ..
240356     1
182178     1
184005     1
209022     1
136610     1
Name: fnlwgt, Length: 28523, dtype: int64
HS-grad         15784
Some-college    10878
Bachelors        8025
Masters          2657
Assoc-voc        2061
11th             1812
Assoc-acdm       1601
10th             1389
7th-8th           955
Prof-school       834
9th               756
12th              657
Doctorate         594
5th-6th           509
1st-4th           247
Preschool          83
Name: education, dtype: int64
9     15784


In [8]:
df.workclass.value_counts()

Private             33906
Self-emp-not-inc     3862
Local-gov            3136
?                    2799
State-gov            1981
Self-emp-inc         1695
Federal-gov          1432
Without-pay            21
Never-worked           10
Name: workclass, dtype: int64

In [9]:
# replace '?' values in workclass variable with `NaN`


df['workclass'].replace('?', np.NaN, inplace=True)

In [10]:
# again check the frequency distribution of values in workclass variable

df.workclass.value_counts()

Private             33906
Self-emp-not-inc     3862
Local-gov            3136
State-gov            1981
Self-emp-inc         1695
Federal-gov          1432
Without-pay            21
Never-worked           10
Name: workclass, dtype: int64

In [11]:
# check frequency distribution of values in occupation variable

df.occupation.value_counts()

Prof-specialty       6172
Craft-repair         6112
Exec-managerial      6086
Adm-clerical         5611
Sales                5504
Other-service        4923
Machine-op-inspct    3022
?                    2809
Transport-moving     2355
Handlers-cleaners    2072
Farming-fishing      1490
Tech-support         1446
Protective-serv       983
Priv-house-serv       242
Armed-Forces           15
Name: occupation, dtype: int64

In [12]:
# replace '?' values in occupation variable with `NaN`

df['occupation'].replace('?', np.NaN, inplace=True)

In [13]:
# again check the frequency distribution of values in occupation variable

df.occupation.value_counts()

Prof-specialty       6172
Craft-repair         6112
Exec-managerial      6086
Adm-clerical         5611
Sales                5504
Other-service        4923
Machine-op-inspct    3022
Transport-moving     2355
Handlers-cleaners    2072
Farming-fishing      1490
Tech-support         1446
Protective-serv       983
Priv-house-serv       242
Armed-Forces           15
Name: occupation, dtype: int64

In [14]:
# check frequency distribution of values in native_country variable

df.native_country.value_counts()

United-States                 43832
Mexico                          951
?                               857
Philippines                     295
Germany                         206
Puerto-Rico                     184
Canada                          182
El-Salvador                     155
India                           151
Cuba                            138
England                         127
China                           122
South                           115
Jamaica                         106
Italy                           105
Dominican-Republic              103
Japan                            92
Guatemala                        88
Poland                           87
Vietnam                          86
Columbia                         85
Haiti                            75
Portugal                         67
Taiwan                           65
Iran                             59
Greece                           49
Nicaragua                        49
Peru                        

In [15]:
# replace '?' values in native_country variable with `NaN`

df['native_country'].replace('?', np.NaN, inplace=True)

In [16]:
# again check the frequency distribution of values in native_country variable

df.native_country.value_counts()

United-States                 43832
Mexico                          951
Philippines                     295
Germany                         206
Puerto-Rico                     184
Canada                          182
El-Salvador                     155
India                           151
Cuba                            138
England                         127
China                           122
South                           115
Jamaica                         106
Italy                           105
Dominican-Republic              103
Japan                            92
Guatemala                        88
Poland                           87
Vietnam                          86
Columbia                         85
Haiti                            75
Portugal                         67
Taiwan                           65
Iran                             59
Nicaragua                        49
Greece                           49
Peru                             46
Ecuador                     

In [17]:
df[categorical].isnull().sum()

age                  0
workclass         2799
fnlwgt               0
education            0
education_num        0
marital_status       0
occupation        2809
relationship         0
race                 0
sex                  0
capital_gain         0
capital_loss         0
hours_per_week       0
native_country     857
income               0
dtype: int64

In [18]:
# find numerical variables

numerical = [var for var in df.columns if df[var].dtype!='O']

print('There are {} numerical variables\n'.format(len(numerical)))

print('The numerical variables are :', numerical)

There are 0 numerical variables

The numerical variables are : []


In [19]:
X = df.drop(['income'], axis=1)

y = df['income']

In [20]:
# split X and y into training and testing sets

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 0)

In [21]:
# check the shape of X_train and X_test

X_train.shape, X_test.shape

((34189, 14), (14653, 14))

In [22]:
# display categorical variables

categorical = [col for col in X_train.columns if X_train[col].dtypes == 'O']

categorical

['age',
 'workclass',
 'fnlwgt',
 'education',
 'education_num',
 'marital_status',
 'occupation',
 'relationship',
 'race',
 'sex',
 'capital_gain',
 'capital_loss',
 'hours_per_week',
 'native_country']

In [23]:
# print percentage of missing values in the categorical variables in training set

X_train[categorical].isnull().mean()

age               0.000000
workclass         0.057211
fnlwgt            0.000000
education         0.000000
education_num     0.000000
marital_status    0.000000
occupation        0.057358
relationship      0.000000
race              0.000000
sex               0.000000
capital_gain      0.000000
capital_loss      0.000000
hours_per_week    0.000000
native_country    0.017608
dtype: float64

In [24]:
# impute missing categorical variables with most frequent value

for df2 in [X_train, X_test]:
    df2['workclass'].fillna(X_train['workclass'].mode()[0], inplace=True)
    df2['occupation'].fillna(X_train['occupation'].mode()[0], inplace=True)
    df2['native_country'].fillna(X_train['native_country'].mode()[0], inplace=True) 

/Users/mohannollu/opt/anaconda3/lib/python3.8/site-packages/pandas/core/series.py:4463: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


In [25]:
# check missing values in categorical variables in X_train

X_train[categorical].isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
dtype: int64

In [26]:
# check missing values in X_train

X_train.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
dtype: int64

In [27]:
X_train[categorical].head()

age         workclass  fnlwgt     education education_num  \
3834   27         Local-gov  131310  Some-college            10   
34744  42           Private  367533          10th             6   
2023   32  Self-emp-not-inc   38158     Assoc-voc            11   
1581   36           Private  148903  Some-college            10   
4613   29           Private  359155       HS-grad             9   

           marital_status        occupation   relationship   race     sex  \
3834   Married-civ-spouse      Adm-clerical      Own-child  White  Female   
34744  Married-civ-spouse      Craft-repair      Own-child  Other    Male   
2023   Married-civ-spouse   Farming-fishing        Husband  White    Male   
1581        Never-married   Exec-managerial  Not-in-family  White  Female   
4613            Separated  Transport-moving      Unmarried  White  Female   

      capital_gain capital_loss hours_per_week native_country  
3834             0            0             40  United-States  
34744            0            0             43  United-States  
2023          7298            0             70  United-States  
1581             0            0             60  United-States  
4613             0            0             30  United-States

In [28]:
# import category encoders

import category_encoders as ce

In [29]:
# encode remaining variables with one-hot encoding

encoder = ce.OneHotEncoder(cols=['workclass', 'education', 'marital_status', 'occupation', 'relationship', 
                                 'race', 'sex', 'native_country'])

X_train = encoder.fit_transform(X_train)

X_test = encoder.transform(X_test)

/Users/mohannollu/opt/anaconda3/lib/python3.8/site-packages/category_encoders/utils.py:21: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead
  elif pd.api.types.is_categorical(cols):


In [30]:
X_train.head()

age  workclass_1  workclass_2  workclass_3  workclass_4  workclass_5  \
3834   27            1            0            0            0            0   
34744  42            0            1            0            0            0   
2023   32            0            0            1            0            0   
1581   36            0            1            0            0            0   
4613   29            0            1            0            0            0   

       workclass_6  workclass_7  workclass_8  fnlwgt  ...  native_country_32  \
3834             0            0            0  131310  ...                  0   
34744            0            0            0  367533  ...                  0   
2023             0            0            0   38158  ...                  0   
1581             0            0            0  148903  ...                  0   
4613             0            0            0  359155  ...                  0   

       native_country_33  native_country_34  native_country_35  \
3834                   0                  0                  0   
34744                  0                  0                  0   
2023                   0                  0                  0   
1581                   0                  0                  0   
4613                   0                  0                  0   

       native_country_36  native_country_37  native_country_38  \
3834                   0                  0                  0   
34744                  0                  0                  0   
2023                   0                  0                  0   
1581                   0                  0                  0   
4613                   0                  0                  0   

       native_country_39  native_country_40  native_country_41  
3834                   0                  0                  0  
34744                  0                  0                  0  
2023                   0                  0                  0  
1581                   0                  0                  0  
4613                   0                  0                  0  

[5 rows x 105 columns]

In [31]:
X_test.head()

age  workclass_1  workclass_2  workclass_3  workclass_4  workclass_5  \
38114  36            0            1            0            0            0   
39215  57            0            0            0            0            1   
44249  49            0            1            0            0            0   
10284  54            0            0            1            0            0   
26725  17            0            0            1            0            0   

       workclass_6  workclass_7  workclass_8  fnlwgt  ...  native_country_32  \
38114            0            0            0   65624  ...                  0   
39215            0            0            0   37345  ...                  0   
44249            0            0            0  287647  ...                  0   
10284            0            0            0  183668  ...                  0   
26725            0            0            0  228786  ...                  0   

       native_country_33  native_country_34  native_country_35  \
38114                  0                  0                  0   
39215                  0                  0                  0   
44249                  0                  0                  0   
10284                  0                  0                  0   
26725                  0                  0                  0   

       native_country_36  native_country_37  native_country_38  \
38114                  0                  0                  0   
39215                  0                  0                  0   
44249                  0                  0                  0   
10284                  0                  0                  0   
26725                  0                  0                  0   

       native_country_39  native_country_40  native_country_41  
38114                  0                  0                  0  
39215                  0                  0                  0  
44249                  0                  0                  0  
10284                  0                  0                  0  
26725                  0                  0                  0  

[5 rows x 105 columns]

In [32]:
cols = X_train.columns

In [33]:
from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()

X_train = scaler.fit_transform(X_train)

X_test = scaler.transform(X_test)

In [34]:
X_test = pd.DataFrame(X_test, columns=[cols])

In [35]:
X_train = pd.DataFrame(X_train, columns=[cols])

In [36]:
X_train.head()

age workclass_1 workclass_2 workclass_3 workclass_4 workclass_5  \
0 -0.50         1.0        -1.0         0.0         0.0         0.0   
1  0.25         0.0         0.0         0.0         0.0         0.0   
2 -0.25         0.0        -1.0         1.0         0.0         0.0   
3 -0.05         0.0         0.0         0.0         0.0         0.0   
4 -0.40         0.0         0.0         0.0         0.0         0.0   

  workclass_6 workclass_7 workclass_8    fnlwgt  ... native_country_32  \
0         0.0         0.0         0.0 -0.387237  ...               0.0   
1         0.0         0.0         0.0  1.575857  ...               0.0   
2         0.0         0.0         0.0 -1.161362  ...               0.0   
3         0.0         0.0         0.0 -0.241033  ...               0.0   
4         0.0         0.0         0.0  1.506233  ...               0.0   

  native_country_33 native_country_34 native_country_35 native_country_36  \
0               0.0               0.0               0.0               0.0   
1               0.0               0.0               0.0               0.0   
2               0.0               0.0               0.0               0.0   
3               0.0               0.0               0.0               0.0   
4               0.0               0.0               0.0               0.0   

  native_country_37 native_country_38 native_country_39 native_country_40  \
0               0.0               0.0               0.0               0.0   
1               0.0               0.0               0.0               0.0   
2               0.0               0.0               0.0               0.0   
3               0.0               0.0               0.0               0.0   
4               0.0               0.0               0.0               0.0   

  native_country_41  
0               0.0  
1               0.0  
2               0.0  
3               0.0  
4               0.0  

[5 rows x 105 columns]

# Model Training

In [37]:
# train a Gaussian Naive Bayes classifier on the training set
from sklearn.naive_bayes import GaussianNB


# instantiate the model
gnb = GaussianNB()


# fit the model
gnb.fit(X_train, y_train)

GaussianNB()

In [38]:
y_pred = gnb.predict(X_test)

y_pred

array(['<=50K', '>50K', '>50K', ..., '<=50K', '<=50K', '<=50K'],
      dtype='<U5')

In [39]:
from sklearn.metrics import accuracy_score

print('Model accuracy score: {0:0.4f}'. format(accuracy_score(y_test, y_pred)))

Model accuracy score: 0.8057


In [40]:
y_pred_train = gnb.predict(X_train)

y_pred_train

array(['<=50K', '<=50K', '>50K', ..., '>50K', '<=50K', '>50K'],
      dtype='<U5')

In [41]:
# print the scores on training and test set

print('Training set score: {:.4f}'.format(gnb.score(X_train, y_train)))

print('Test set score: {:.4f}'.format(gnb.score(X_test, y_test)))

Training set score: 0.8072
Test set score: 0.8057


#### The training-set accuracy score is 0.8072 while the test-set accuracy to be 0.8057. These two values are quite comparable. So, there is no sign of overfitting.